File: ublong-agents/app/models/case.py
from pydantic import BaseModel, Field, ConfigDict, AliasChoices
from typing import Optional, Any


class CaseInput(BaseModel):
    # Accept both the agents-native field names and the backend/frontend canonical
    # names (country_of_birth, date_of_birth, current_country). populate_by_name
    # lets internal code keep using the descriptive names while the Node backend
    # may POST snake_case "country_of_birth"-style payloads.
    model_config = ConfigDict(populate_by_name=True, extra="ignore")

    child_birth_country: str = Field(
        validation_alias=AliasChoices("child_birth_country", "country_of_birth")
    )
    child_birth_date: str = Field(
        validation_alias=AliasChoices("child_birth_date", "date_of_birth")
    )
    child_current_country: str = Field(
        validation_alias=AliasChoices("child_current_country", "current_country")
    )
    father_nationality: str = ""
    mother_nationality: str = ""
    parents_documents_held: list[str] = Field(default_factory=list)
    child_age_months: int = 0
    additional_context: str = ""


class AgentEvent(BaseModel):
    event_type: str  # "thinking" | "result" | "error"
    agent_name: str
    content: str
    timestamp: str


class CaseResult(BaseModel):
    case_id: str
    legal_pathway: str
    required_documents: list[str]
    available_substitutes: list[str]
    missing_documents: list[str]
    submission_office: str
    estimated_timeline: str
    cover_letter_draft: str
    confidence_score: float
    # Array of flags/notes — matches the backend Mongo schema (country_specific_notes: [String]).
    country_specific_notes: list[str] = Field(default_factory=list)
    # Per-agent summaries surfaced for the dashboard / case record (optional).
    intake_summary: Optional[dict[str, Any]] = None
    research_summary: Optional[dict[str, Any]] = None
    gap_summary: Optional[dict[str, Any]] = None

File: ublong-backend/src/models/Case.js
const mongoose = require("mongoose");

const intakeSchema = new mongoose.Schema(
  {
    // PRD CaseInput fields
    country_of_birth: { type: String, required: true },
    date_of_birth: { type: String, required: true },
    current_country: { type: String, required: true },
    father_nationality: { type: String, default: "" },
    mother_nationality: { type: String, default: "" },
    parents_documents_held: { type: [String], default: [] },
    child_age_months: { type: Number },
    additional_context: { type: String, default: "" },

    // Frontend-extended fields
    child_name: { type: String, default: "" },
    gender: { type: String, default: "" },
    father_name: { type: String, default: "" },
    mother_name: { type: String, default: "" },
    current_location: { type: String, default: "" },
  },
  { _id: false }
);

const resultSchema = new mongoose.Schema(
  {
    case_id: String,
    legal_pathway: String,
    required_documents: [String],
    available_substitutes: [String],
    missing_documents: [String],
    submission_office: String,
    estimated_timeline: String,
    cover_letter_draft: String,
    confidence_score: Number,
    country_specific_notes: [String],
    intake_summary: mongoose.Schema.Types.Mixed,
    research_summary: mongoose.Schema.Types.Mixed,
    gap_summary: mongoose.Schema.Types.Mixed,
  },
  { _id: false }
);

const caseSchema = new mongoose.Schema(
  {
    case_id: { type: String, unique: true, sparse: true },
    created_by: { type: mongoose.Schema.Types.ObjectId, ref: "User", required: true },
    status: {
      type: String,
      enum: ["pending", "processing", "completed", "flagged_for_review"],
      default: "pending",
    },
    intake: { type: intakeSchema, required: true },
    result: { type: resultSchema, default: null },
  },
  { timestamps: { createdAt: "created_at", updatedAt: "updated_at" } }
);

module.exports = mongoose.model("Case", caseSchema);
